http-server/src/server.rs

```diff
@@ -153,6 +157,24 @@ impl Server {
 				Ok::<_, HyperError>(service_fn(move |request| {
 					let methods = methods.clone();
 					let access_control = access_control.clone();
+
+					let execute =
```

Comment:
I know, and I agree, and I tried but couldn't make it work; I'd have to look up the `method` from the `methods` hash map in `start()` and pass it along so it didn't really shorten up the code much.

Lately I've started to shift my views on when it's right to refactor for briefness. The rule of thumb is still "terse is good", but for some cases I've started to think it's alright to keep the code long when it's "the main loop", whatever that means for each specific case, i.e. the most important thing that a given program does. Splitting things up too much can force the reader to jump around more than is ideal and even if it's long it is sometimes more readable to keep it all in one place. I know this is hand wavey (and for this case here I wanted to do exactly what you suggest), but yeah, just wanted to share those thoughts.

Reply:
It's mainly because of the closure that is passed to hyper but yeah this changes made the code really quite hard to read, it wasn't great before either.

Maybe we split it the response handling to helper functions, basically you have to read the bytes here to take ownership over it which used to later to build a message to send to the background task.

Reply:
Personally, I found [this](https://github.com/paritytech/jsonrpsee/blob/master/ws-client/src/client.rs#L554-#L599) more readable but it's still quite long ^^

(maybe because I wrote it lol)

Reply:
K, I tried another approach, moving `execute` to be a method on `Server` so I can access `self.root` (i.e. the function pointers we call to run the request), but it doesn't work because `self` is moved into the closure when calling `make_service_fn()`.
On the client side it's a bit easier because we don't have to call anything. Or I'm too limited to see how to do it! Pointers welcome!

Reply:
We could just add another parameter, `&Methods` to the execute function so execute becomes:

```rust
fn execute(methods: &Methods, id: Option<&RawValue>, tx: RpcSender, method_name: &str, params: Option<&RawValue>) {
	if let Some(method) = methods.get(method_name) {
		let params = RpcParams::new(params.map(|params| params.get()));
		// NOTE(niklasad1): connection ID is unused thus hardcoded to `0`.
		if let Err(err) = (method)(id, params, &tx, 0) {
			log::error!("execution of method call {} failed: {:?}, request id={:?}", method_name, err, id);
		}
	} else {
		send_error(id, tx, JsonRpcErrorCode::MethodNotFound.into());
	}
}
```

and then call it like `execute(&methods, id, &tx, &method_name, params);`

I guess the tradeoff is having a long function signature for `execute`, though.

I think I agree with your analysis of `main_loops` though. Often it can be hard to make them shorter and would result in more confusing code than just keeping it all together. In this case I think `execute` fn is self-explanatory enough in that it just executes the right function for the rpc call.

Reply:
@insipx I could have sworn I tried that; your version works too. I can go either way here, @niklasad1 thoughts?

Reply:
FWIW the performance of @insipx's version is identical.